src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -157,15 +256,293 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+                            {
+                                isDataLeft = true;
+                                srcTensorGetters[i].BufferTrainingData();
+                            }
+
+                            if ((rows % args.BatchSize) == 0)
+                            {
+                                isDataLeft = false;
+                                (loss, metric) = TrainBatch(inputColIndices, inputsForTraining, srcTensorGetters, fetchList, args);
+                            }
+                        }
+                        if (isDataLeft)
+                        {
+                            isDataLeft = false;
+                            ch.Warning("Not training on the last batch. The batch size is less than {0}.", args.BatchSize);
+                        }
+                        ch.Info("Loss: {0}, Metric: {1}", fetchList.Count > 0 ? loss.ToString("#.####") : "None",
+                                fetchList.Count > 1 ? metric.ToString("#.####") : "None");
+
+                        ch.Done();
+                    }
+                }
+            }
+            UpdateModelOnDisk(model, args);
+        }
+
+        private (float loss, float metric) TrainBatch(int[] inputColIndices,
+            string[] inputsForTraining,
+            ITensorValueGetter[] srcTensorGetters,
+            List<string> fetchList,
+            Arguments args)
+        {
+            float loss = 0;
+            float metric = 0;
+            var runner = Session.GetRunner();
+            for (int i = 0; i < inputColIndices.Length; i++)
+            {
+                var inputName = inputsForTraining[i];
+                runner.AddInput(inputName, srcTensorGetters[i].GetBufferedBatchTensor());
+            }
+            if (args.LearningRateOperation != null)
+                runner.AddInput(args.LearningRateOperation, new TFTensor(args.LearningRate));
+            runner.AddTarget(args.OptimizationOperation);
+
+            if (fetchList.Count > 0)
+                runner.Fetch(fetchList.ToArray());
+
+            var tensor = runner.Run();
+            loss = tensor.Length > 0 ? (float)tensor[0].GetValue() : 0.0f;
+            metric = tensor.Length > 1 ? (float)tensor[1].GetValue() : 0.0f;
+
+            return (loss, metric);
+        }
+
+        /// <summary>
+        /// Updates the model on the disk.
+        /// After retraining Session and Graphs are both up-to-date
+        /// However model on disk is not which is used to serialzed to ML.Net stream
+        /// </summary>
+        private void UpdateModelOnDisk(string modelDir, Arguments args)
+        {
+            try
+            {
+                // Save the model on disk
+                var path = Path.Combine(modelDir, DefaultModelFileNames.TmpMlnetModel);
+                Session.GetRunner().AddInput(args.SaveLocationOperation, TFTensor.CreateString(Encoding.UTF8.GetBytes(path)))
+                        .AddTarget(args.SaveOperation).Run();
```

Comment:
>AddTarget [](start = 25, length = 9)

Why do we use AddTarget here instead of Fetch ?

Reply:
Fetch is for retrieving some output. The operation `SaveOperation` technically does not get back any output from the graph. So the operation should be executed as target.

---
In reply to: [221799044](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221799044) [](ancestors = 221799044)